### ISSUE TITLE
Not able to connect with SignalR hub using websockets transport type

### ISSUE THREAD

**Issue by njangir2** [aspnet/SignalR#1607]
I am not able to connect with the SignalRHub from my .net core console application.
Strange thing is that from .net framework console application it is working fine.
I am using latest SignalR library "Microsoft.AspNetCore.SignalR.Client" Version="1.0.0-preview1-final".
Error that I am getting is given below below-
```
info: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[1]
      Starting transport. Transfer mode: Text.
fail: Microsoft.AspNetCore.Sockets.Client.HttpConnection[10]
      Failed to start connection. Error starting transport 'WebSocketsTransport'.
System.Net.WebSockets.WebSocketException (0x80004005): Unable to connect to the remote server ---> System.Net.WebSockets.WebSocketException (0x80004005): Unable to connect to the remote server ---> System.Net.Http.WinHttpException (0x80072F8F): A security error occurred
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at System.Net.WebSockets.WinHttpWebSocket.<ConnectAsync>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.WebSocketHandle.<ConnectAsyncCore>d__20.MoveNext()
   at System.Net.WebSockets.WebSocketHandle.<ConnectAsyncCore>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.ClientWebSocket.<ConnectAsyncCore>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport.<Connect>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport.<StartAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Sockets.Client.HttpConnection.<StartTransport>d__47.MoveNext()
```

Code that I am using to connect with the SignalRHub is given below-
```
public async Task<bool> ConnectToSignalRService(string hubUrl, string hubName)
{
	try
	{

		Uri.TryCreate(new Uri(hubUrl), hubName, out Uri url);

		HttpClientHandler handler = new HttpClientHandler
		{
			ClientCertificateOptions = ClientCertificateOption.Manual,
			ServerCertificateCustomValidationCallback = (httpRequestMessage, cert, cetChain, policyErrors) => true,
			SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls
		};

		Connection = new HubConnectionBuilder()
		.WithUrl(url)
		.WithTransport(transportType)
		.WithMessageHandler(handler)
		.WithConsoleLogger()
		.Build();
		
		Connection.On<string>("GetSignalRConnectionUniqueId",
		uniqueId =>
		{
			//DO something here
		});

		//Added conditions because the way to ignore/accept self-signed certificate using .NET Framework and using .NET Standard 2.0 it is different
#if NET461
System.Net.ServicePointManager.ServerCertificateValidationCallback += (o, c, ch, er) => true;                           
#endif
		await Connection.StartAsync();
		await Connection.InvokeAsync("SetConnection");
	}
	catch (Exception ex)
	{
		throw;
	}
	return true;
}
```
The problem is quite strange to me as same code is working in .net framework console application but not working in .net core console application.
Please let me know if I am missing something here.

**Comment by anurse** [comment]
@Tratcher does anything look awry here as far as certificate validation parameters?

@njangir2 Can you specify the operating system you are running the .NET Core application on? It looks like Windows, but I'd like to confirm that.

**Comment by anurse** [comment]
Hrm, actually. The issue seems to be coming from WebSockets, which is due to a bug in the ClientWebSocket class in corefx: https://github.com/dotnet/corefx/issues/12038

Are you using .NET Core 2.1-preview1 or 2.0?

**Comment by njangir2** [comment]
@anurse: My client application is running on Windows 10.
I am using .NET Core 2.0 in the SignalR client console application.
My stateless service fabric application (Having SignalR hub) deployed in the service fabric cluster by using a self signed certificate.
SignalR hub side the library that I am using is `Microsoft.AspNetCore.SignalR" Version="1.0.0-preview1-final"`
By using the same SignalR client application code my application run successfully in a .net framework console application using transport type web sockets but didn't work in .net core.
One more point that I want to let you know that when I change transport type to ServerSentEvents or LongPolling then .net core application works smoothly without any issue.
Let me know if you need any other detail here.

**Comment by anurse** [comment]
Yeah, I confirmed that this is an issue with `ClientWebSocket` and not something we can fix in SignalR. A simple WebSockets-only application (no SignalR involved) exhibits the same error.

dotnet/corefx#12038 is tracking adding this functionality to .NET Core 2.1, but since it requires a new API (for the cert validation callback) they won't be patching 2.0 with a fix. I suggest you try .NET Core 2.1-preview1 and prepare to upgrade your application to .NET Core 2.1 when it releases in the first half of this year. Since SignalR is part of .NET Core 2.1-preview1, upgrading will be a good idea anyway.

If you absolutely have to use .NET Core 2.0, you'll need to use a properly trusted certificate (either by getting a real certificate, using a hosting provider like Azure Web Apps, or installing your certificate into the Client's trusted certificates list).

I'm closing this as there's nothing SignalR can do to fix this issue. It can continue to be tracked in the corefx issue

**Comment by njangir2** [comment]
@anurse, I have updated .net core for my console app to the "2.1.0-preview1-26216-03" but I am still facing the same issue.
dotnet/corefx#12038 is still in open state, do I need to wait for the closing of this issue?
The self signed certificate that is being used in my SignalR hub service is already added as a trusted certificate on my local machine.
Please let me know what else I can do here.

**Comment by anurse** [comment]
> The self signed certificate that is being used in my SignalR hub service is already added as a trusted certificate on my local machine.

Then why do you have `ServerCertificateCustomValidationCallback = (httpRequestMessage, cert, cetChain, policyErrors) => true,` in your HttpClientHandler options? If you need that to connect HttpClient, your certificate isn't properly marked as trusted. Is it just there because you forgot to remove it?

> dotnet/corefx#12038 is still in open state, do I need to wait for the closing of this issue?

Correct, the latest versions of 2.1 do not have the fix. My point was just that the fix will only be available in 2.1, it will not be ported to 2.0.
